components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/dao/JDBCSubscriberDAOImpl.java

```diff
@@ -0,0 +1,372 @@
+import static org.wso2.carbon.identity.entitlement.dao.DAOConstants.SQLQueries.UPDATE_SUBSCRIBER_PROPERTIES_SQL;
+
+public class JDBCSubscriberDAOImpl implements SubscriberDAO {
+
+    private static final Log LOG = LogFactory.getLog(JDBCSubscriberDAOImpl.class);
+    private static final String ERROR_SUBSCRIBER_ID_NULL = "Subscriber Id can not be null";
+
+    /**
+     * Gets the requested subscriber.
+     *
+     * @param subscriberId  subscriber ID.
+     * @param returnSecrets whether the subscriber should get returned with secret(decrypted) values or not.
+     * @return publisher data holder.
+     * @throws EntitlementException If an error occurs.
+     */
+    @Override
+    public PublisherDataHolder getSubscriber(String subscriberId, boolean returnSecrets) throws EntitlementException {
+
+        int tenantId = CarbonContext.getThreadLocalCarbonContext().getTenantId();
+
+        try (Connection connection = IdentityDatabaseUtil.getDBConnection(false)) {
+            try (NamedPreparedStatement preparedStmt = new NamedPreparedStatement(connection, GET_SUBSCRIBER_SQL)) {
+                preparedStmt.setString(SUBSCRIBER_ID, subscriberId);
+                preparedStmt.setInt(TENANT_ID, tenantId);
+                try (ResultSet resultSet = preparedStmt.executeQuery()) {
+                    if (resultSet.next()) {
+                        return getPublisherDataHolder(resultSet, returnSecrets);
+                    }
+                    return null;
+                }
+            }
+        } catch (SQLException e) {
+            throw new EntitlementException(String.format("Error while retrieving subscriber details of id : %s",
+                    subscriberId), e);
+        }
+    }
+
+    /**
+     * Gets all subscriber IDs.
+     *
+     * @param filter search string.
+     * @return list of subscriber IDs.
+     * @throws EntitlementException If an error occurs.
+     */
+    @Override
+    public List<String> listSubscriberIds(String filter) throws EntitlementException {
+
+        int tenantId = CarbonContext.getThreadLocalCarbonContext().getTenantId();
+        List<String> subscriberIdList = new ArrayList<>();
+
+        try (Connection connection = IdentityDatabaseUtil.getDBConnection(false)) {
+            try (NamedPreparedStatement preparedStmt = new NamedPreparedStatement(connection, GET_SUBSCRIBER_IDS_SQL)) {
+                preparedStmt.setInt(TENANT_ID, tenantId);
+
+                try (ResultSet subscriberIds = preparedStmt.executeQuery()) {
+                    while (subscriberIds.next()) {
+                        subscriberIdList.add(subscriberIds.getString(SUBSCRIBER_ID));
+                    }
+                }
+            }
+        } catch (SQLException e) {
+            throw new EntitlementException("Error while retrieving subscriber ids", e);
+        }
+
+        return EntitlementUtil.filterSubscribers(subscriberIdList, filter);
```

Comment:
Can we do the filtering in the db side itself?

Reply:
This filter string supports cases which cannot be easily supported with DB queries.
eg: wildcards - user*, specific characters - user[123], etc.

Reply:
lets indicate that with a comment here..